src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -1807,6 +1877,160 @@ public override TypeKind TypeKind
+        {
+            // PROTOTYPE we should also decode `implicit` vs. `explicit`
+            // PROTOTYPE consider optimizing/caching
+
+            TypeSymbol? foundUnderlyingType;
+            if (!tryDecodeExtensionType(out foundUnderlyingType, out baseExtensions))
+            {
+                if (foundUnderlyingType is null)
+                {
+                    foundUnderlyingType = new UnsupportedMetadataTypeSymbol();
+                }
+
+                if (baseExtensions.IsDefault)
+                {
+                    baseExtensions = ImmutableArray<NamedTypeSymbol>.Empty;
+                }
+            }
+
+            underlyingType = foundUnderlyingType;
+
+            return;
+
+            bool tryDecodeExtensionType([NotNullWhen(true)] out TypeSymbol? underlyingType, out ImmutableArray<NamedTypeSymbol> baseExtensions)
+            {
+                var found = IsExtensionType(out MethodDefinitionHandle? foundMarkerMethod);
+                Debug.Assert(found);
+                var markerMethod = foundMarkerMethod!.Value;
+                var moduleSymbol = this.ContainingPEModule;
+
+                underlyingType = null;
+                baseExtensions = default;
+
+                MethodAttributes localFlags;
+                moduleSymbol.Module.GetMethodDefPropsOrThrow(markerMethod, name: out _, implFlags: out _, out localFlags, rva: out _);
+                // PROTOTYPE do we want to tighten the flags check further?
+                if ((localFlags & MethodAttributes.Private) == 0 ||
+                    (localFlags & MethodAttributes.Static) == 0)
+                {
+                    return false;
+                }
+
+                SignatureHeader signatureHeader;
+                BadImageFormatException mrEx;
+                ParamInfo<TypeSymbol>[] paramInfos = new MetadataDecoder(moduleSymbol, context: this)
+                    .GetSignatureForMethod(markerMethod, out signatureHeader, out mrEx);
+
+                if (mrEx is not null || signatureHeader.IsGeneric || paramInfos.Length <= 1)
+                {
+                    return false;
+                }
+
+                var baseExtensionsBuilder = ArrayBuilder<NamedTypeSymbol>.GetInstance(paramInfos.Length - 1);
+                for (int i = 0; i < paramInfos.Length; i++)
+                {
+                    var paramInfo = paramInfos[i];
+                    TypeSymbol type = paramInfo.Type;
+
+                    if (paramInfo.IsByRef || !paramInfo.CustomModifiers.IsDefault)
+                    {
+                        var info = new CSDiagnosticInfo(ErrorCode.ERR_MalformedExtensionInMetadata, Name); // PROTOTYPE need to report use-site diagnostic
+                        type = new ExtendedErrorTypeSymbol(type, LookupResultKind.NotReferencable, info, unreported: true);
+                    }
+
+                    type = ApplyTransforms(type, paramInfo.Handle, moduleSymbol);
```

Comment:
>type = ApplyTransforms(type, paramInfo.Handle, moduleSymbol); [](http://example.com/codeflow?start=20&length=61)

Transforms around extension type references might be tricky. Consider leaving a PROTOTYPE comment.